Add support for managed app plugins

thepalbi opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana-plugin-sdk-go/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

4. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

-->

**What this PR does / why we need it**:
The SDK introduces to plugin development a way to initialize the backend plugin instances with configurations contained in the plugin's `jsonData` or `securedJsonData`. This is achieved by having _managed_ backend plugin instances.

Currently, just datasource plugins are supported for this managed approach.

This PR is based on the datasource work, and does the same for app plugins, which should have just one plugin instance alive at each moment.

**Which issue(s) this PR fixes**:
Fixes #534

**Special notes for your reviewer**:
- ~~Tests haven't been added yet. Would be happy to add if the implemented approach is accepted.~~

thepalbi committed:
wip

thepalbi committed:
revert datasource instance provider changes

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana-plugin-sdk-go?pullRequest=535) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana-plugin-sdk-go?pullRequest=535) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana-plugin-sdk-go?pullRequest=535) it.</sub>

marefr commented on backend/app/instance_provider.go:
```diff
@@ -0,0 +1,60 @@
+	}
+
+	// Since app plugins have just one instance, use pluginID as instance cache key
+	return pluginContext.PluginID, nil
```
You need to include the `OrgID` as well because there could be one enabled app plugin per organization.

marefr commented on backend/app/instance_provider.go:
```diff
@@ -0,0 +1,60 @@
+// updated in Grafana.
+// Cache key is based on the numerical data source identifier.
+// If fn is nil, NewInstanceProvider panics.
+func NewInstanceProvider(fn InstanceFactoryFunc) instancemgmt.InstanceProvider {
```
Please fix comments above: s/data source/app and adapt to used cache key

thepalbi replied:
Done!

marefr commented on backend/app/instance_provider.go:
```diff
@@ -0,0 +1,60 @@
+// InstanceFactoryFunc factory method for creating data source instances.
+type InstanceFactoryFunc func(settings backend.AppInstanceSettings) (instancemgmt.Instance, error)
+
+// NewInstanceManager creates a new data source instance manager,
```
```suggestion
// NewInstanceManager creates a new app instance manager,
```

marefr commented on backend/app/instance_provider.go:
```diff
@@ -0,0 +1,60 @@
+	"github.com/grafana/grafana-plugin-sdk-go/backend/instancemgmt"
+)
+
+// InstanceFactoryFunc factory method for creating data source instances.
```
```suggestion
// InstanceFactoryFunc factory method for creating app instances.
```

marefr commented on backend/app/manage.go:
```diff
@@ -0,0 +1,44 @@
+	GRPCSettings backend.GRPCSettings
+}
+
+// Manage starts serving the data source over gPRC with automatic instance management.
```
```suggestion
// Manage starts serving the app over gPRC with automatic instance management.
```

marefr reviewed (changes requested):
Looks good in general, but needs some minor changes.

Would be nice if you can add package documentation similar to https://github.com/grafana/grafana-plugin-sdk-go/blob/main/backend/datasource/doc.go as well.

thepalbi committed:
Update backend/app/instance_provider.go

Co-authored-by: Marcus Efraimsson <marcus.efraimsson@gmail.com>

thepalbi committed:
Update backend/app/instance_provider.go

Co-authored-by: Marcus Efraimsson <marcus.efraimsson@gmail.com>

thepalbi committed:
Update backend/app/manage.go

Co-authored-by: Marcus Efraimsson <marcus.efraimsson@gmail.com>

thepalbi committed:
Address PR comments

thepalbi committed:
Added tests

thepalbi committed:
adding docfile

thepalbi commented:
> Looks good in general, but needs some minor changes.
> 
> Would be nice if you can add package documentation similar to https://github.com/grafana/grafana-plugin-sdk-go/blob/main/backend/datasource/doc.go as well.

@marefr thanks for the review 😀! Addressed all comments and added tests for the instance provider.

marefr reviewed (approved):
LGTM
